New feature implemented: UserLocation pins displayed on map indicate their "freshness"/relevant status. UserLocation can be "fresh" or "stale" according to how old the sample is.

BarAgam opened the pull request:
New feature implemented: UserLocation pins displayed on map indicate their "freshness"/relevant status. UserLocation can be "fresh" or "stale" according to how old the sample is.

Added classes:
UserLocation a data-class that encapsulates both the associated user's Id and LocationSample
UsersLocationViewModel - class that holds all the UserLocation objects that needs to be displayed.
    Takes in a UserLocationSymbolizer that given UserLocation determines how to symbolize it over the map.

Refactored:
ViewerFragment - All the logic/responsibilty of displaying user-location is now moved inside the UsersLocationViewModel, therefore the class was refactored to use it.
 It uses a periodical runnable (that is executed via the fragment's thread) that updates the user location pins every 5 seconds, regardless of when the last user location message arrived [this is done to allow the update of pins becoming stale over time].
 It implements a symbolizer named ElapsedTimedUserLocationSymbolizer that creates different symbols to user location pins depending of how "old" the user-location is (if its stale or not).
 Initialized UsersLocationViewModel with said symbolizer.
Boyscout cleaning - turned the anonymous user location new message handler class into an inner-class.
refactored it too to use the UsersLocationViewModel.

Boyscout cleaning - double semicolon from Point class.

barmako committed:
New feature implemented: UserLocation pins displayed on map indicate their "freshness"/relevant status. UserLocation can be "fresh" or "stale" according to how old the sample is.

Added classes:
UserLocation a data-class that encapsulates both the associated user's Id and LocationSample
UsersLocationViewModel - class that holds all the UserLocation objects that needs to be displayed.
    Takes in a UserLocationSymbolizer that given UserLocation determines how to symbolize it over the map.

Refactored:
ViewerFragment - All the logic/responsibilty of displaying user-location is now moved inside the UsersLocationViewModel, therefore the class was refactored to use it.
 It uses a periodical runnable (that is executed via the fragment's thread) that updates the user location pins every 5 seconds, regardless of when the last user location message arrived [this is done to allow the update of pins becoming stale over time].
 It implements a symbolizer named ElapsedTimedUserLocationSymbolizer that creates different symbols to user location pins depending of how "old" the user-location is (if its stale or not).
 Initialized UsersLocationViewModel with said symbolizer.
Boyscout cleaning - turned the anonymous user location new message handler class into an inner-class.
refactored it too to use the UsersLocationViewModel.

Boyscout cleaning - double semicolon from Point class.

review-ninja commented:
<a href="https://app.review.ninja/teamagam/GimelGimel/pull/51" target="_blank"><img src="https://app.review.ninja/52010471/pull/51/badge" alt="ReviewNinja"/>

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -59,12 +65,18 @@
 
+    private static final Logger sLogger = LoggerFactory.create(ViewerFragment.class);
+
+    //todo: config. after merging configuration branch move this configuration to Constants.java
```
why 5 seconds? definitely too much. maybe one minute? maybe 10?
we will decide when merging to dev (and move to Constants)

teamagam1 replied:
nevermind

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/entities/UserLocation.java:
```diff
@@ -0,0 +1,19 @@
+package com.teamagam.gimelgimel.app.model.entities;
+
+public class UserLocation {
+    private final String id;
```
mId !fix

BarAgam replied:
!fixed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/entities/UserLocation.java:
```diff
@@ -0,0 +1,19 @@
+
+public class UserLocation {
+    private final String id;
+    private final LocationSample locationSample;
```
same !fix

BarAgam replied:
!fixed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/model/entities/UserLocation.java:
```diff
@@ -0,0 +1,19 @@
+package com.teamagam.gimelgimel.app.model.entities;
+
```
few words about what is this class

BarAgam replied:
!completed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -59,12 +65,18 @@
         SendGeographicMessageDialog.SendGeographicMessageDialogInterface,
         ShowMessageDialogFragment.ShowMessageDialogFragmentInterface, OnGGMapReadyListener {
 
+    private static final Logger sLogger = LoggerFactory.create(ViewerFragment.class);
```
dont we have sLogger from BaseFragment?

teamagam1 replied:
!fix

BarAgam replied:
!fixed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -376,4 +417,66 @@ private void onCreateGeographicMessage(PointGeometry pointGeometry) {
+        public Symbol symbolize(UserLocation userLocation) {
+            long timestampsDeltaMs = getTimestampsDeltaMs(userLocation);
+
+            sLogger.v(String.format("Symbolizing userLocation with id %s and timestamp delta %s",
```
too much Log

teamagam1 replied:
add log for new users, stale users, etc.

teamagam1 replied:
!fix

BarAgam replied:
I cannot know at that point if its a new user/stale-user.
I did logged the type 
And anyways, its a verbose log. !completed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -376,4 +417,66 @@ private void onCreateGeographicMessage(PointGeometry pointGeometry) {
+            return timestampsDeltaMs < USER_LOCATION_STALE_THRESHOLD_MS;
+        }
+
+        private long getTimestampsDeltaMs(UserLocation userLocation) {
```
getTime_S_tampsDeltaMs

teamagam1 replied:
לטובת קריאות

BarAgam replied:
Change the so that logic is now inside UserLocation (and LocationSample). It has a better naming too :)

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -376,4 +417,66 @@ private void onCreateGeographicMessage(PointGeometry pointGeometry) {
+
+        //TODO: config , add these to the static configuration file after merging configuration feature branch
+        private static final long USER_LOCATION_STALE_THRESHOLD_MS = 60 * 1000;
+        private static final String ACTIVE_USER_LOCATION_PIN_COLOR_CSS_STRING = "#7FFF00";
```
no need for _STRING

BarAgam replied:
!fixed

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -376,4 +417,66 @@ private void onCreateGeographicMessage(PointGeometry pointGeometry) {
+
+    private class UserLocationMessageHandler implements MessageBroadcastReceiver.NewMessageHandler {
+
+        private final Logger mLogger = LoggerFactory.create(
```
sLogger

BarAgam replied:
Its a non-static inner class. It cannot have static fields. the logger isn't static :<

teamagam1 commented on app/src/main/java/com/teamagam/gimelgimel/app/view/fragments/ViewerFragment.java:
```diff
@@ -376,4 +417,66 @@ private void onCreateGeographicMessage(PointGeometry pointGeometry) {
+                ElapsedTimeUserLocationSymbolizer.class);
+
+        //TODO: config , add these to the static configuration file after merging configuration feature branch
+        private static final long USER_LOCATION_STALE_THRESHOLD_MS = 60 * 1000;
```
do you think that we need to remove (transparent) users with bigger threshold for

barmako committed:
Moved UserLocation age computation inside the class.
[Added that age computation to LocationSample, which the UserLocation uses]
Renamed constants to match other constants in class.

Added a brief documentation to UserLocation
Changed it's field names to meet conventions.

barmako committed:
Removed ViewerFragment logger and used inherited logger from BaseFragment.
Removed uninformative logs from symbolizer.
